homeassistant/components/voice_assistant/pipeline.py

```diff
@@ -56,69 +55,161 @@ class Pipeline:
+                    "language": self.language,
                 },
             )
         )
 
-        intent_input = request.intent_input
+    def finish(self):
+        """Emit run finish event."""
+        self.event_callback(
+            PipelineEvent(
+                PipelineEventType.RUN_FINISH,
+            )
+        )
 
-        event_callback(
+    async def recognize_intent(
+        self, intent_input: str, conversation_id: str | None
+    ) -> conversation.ConversationResult:
+        """Run intent recognition portion of pipeline."""
+        self.event_callback(
             PipelineEvent(
                 PipelineEventType.INTENT_START,
                 {
-                    "engine": self.conversation_engine or "default",
+                    "engine": self.pipeline.conversation_engine or "default",
                     "intent_input": intent_input,
                 },
             )
         )
 
         conversation_result = await conversation.async_converse(
-            hass=hass,
+            hass=self.hass,
             text=intent_input,
-            conversation_id=request.conversation_id,
-            context=context,
-            language=language,
-            agent_id=self.conversation_engine,
+            conversation_id=conversation_id,
+            context=self.context,
+            language=self.language,
+            agent_id=self.pipeline.conversation_engine,
         )
 
-        event_callback(
+        self.event_callback(
             PipelineEvent(
                 PipelineEventType.INTENT_FINISH,
                 {"intent_output": conversation_result.as_dict()},
             )
         )
 
-        event_callback(
+        return conversation_result
+
+    async def text_to_speech(self, tts_input: str) -> str:
+        """Run text to speech portion of pipeline. Returns URL of TTS audio."""
+        self.event_callback(
             PipelineEvent(
-                PipelineEventType.RUN_FINISH,
+                PipelineEventType.TTS_START,
+                {
+                    "engine": self.pipeline.tts_engine or "default",
+                    "tts_input": tts_input,
+                },
             )
         )
+
+        tts_media = await async_resolve_media(
+            self.hass,
+            tts_generate_media_source_id(
+                self.hass,
+                tts_input,
+                engine=self.pipeline.tts_engine,
+            ),
+        )
+        tts_url = tts_media.url
+
+        self.event_callback(
+            PipelineEvent(
+                PipelineEventType.TTS_FINISH,
+                {"tts_output": tts_url},
+            )
+        )
+
+        return tts_url
+
+
+@dataclass
+class PipelineRequest(ABC):
+    """Request to for a pipeline run."""
+
+    async def execute(
+        self, run: PipelineRun, timeout: int | float | None = DEFAULT_TIMEOUT
+    ):
+        """Run pipeline with optional timeout."""
+        await asyncio.wait_for(
```

Comment:
Can we use `async_timeout` instead? It has less overhead than `asyncio.wait_for`.